From 5163edd1dd648bcec979fa424b3127db9468686d Mon Sep 17 00:00:00 2001
From: Zach Kuznia <zork@chromium.org>
Date: Thu, 1 Jul 2010 14:08:00 +0900
Subject: [PATCH 5/7] Add api to ibus for retreiving unused config values.

This will be used by our autotest scripts to generate an error when input method config values change.

BUG=chromium-os:3672
TEST=none

Review URL: http://codereview.chromium.org/2860022

Ported this change to ibus-1.4 by Peng Huang & Yusuke Sato
---
 gconf/config.c          |   16 ++++++++
 ibus/config.py          |   12 ++++++
 memconf/config.c        |   88 ++++++++++++++++++++++++++++++++++++++++------
 src/ibusconfig.c        |   34 ++++++++++++++++++
 src/ibusconfig.h        |   16 ++++++++-
 src/ibusconfigservice.c |   41 ++++++++++++++++++++++
 src/ibusconfigservice.h |    7 +++-
 7 files changed, 199 insertions(+), 15 deletions(-)

diff --git a/gconf/config.c b/gconf/config.c
index df1976c..15325ff 100644
--- a/gconf/config.c
+++ b/gconf/config.c
@@ -34,6 +34,10 @@ static gboolean     ibus_config_gconf_unset_value   (IBusConfigService      *con
                                                      const gchar            *section,
                                                      const gchar            *name,
                                                      GError                **error);
+static gboolean     ibus_config_gconf_get_unused    (IBusConfigService      *config,
+                                                     GVariant              **unread,
+                                                     GVariant              **unwritten,
+                                                     GError                **error);
 static GConfValue   *_to_gconf_value                (GVariant               *value);
 static GVariant     *_from_gconf_value              (const GConfValue       *gvalue);
 
@@ -48,6 +52,7 @@ ibus_config_gconf_class_init (IBusConfigGConfClass *class)
     IBUS_CONFIG_SERVICE_CLASS (object_class)->set_value   = ibus_config_gconf_set_value;
     IBUS_CONFIG_SERVICE_CLASS (object_class)->get_value   = ibus_config_gconf_get_value;
     IBUS_CONFIG_SERVICE_CLASS (object_class)->unset_value = ibus_config_gconf_unset_value;
+    IBUS_CONFIG_SERVICE_CLASS (object_class)->get_unused = ibus_config_gconf_get_unused;
 }
 
 static void
@@ -300,6 +305,17 @@ ibus_config_gconf_unset_value (IBusConfigService      *config,
     return TRUE;
 }
 
+static gboolean
+ibus_config_gconf_get_unused (IBusConfigService      *config,
+                              GVariant              **unread,
+                              GVariant              **unwritten,
+                              GError                **error)
+{
+    *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                          "Not implemented");
+    return FALSE;
+}
+
 IBusConfigGConf *
 ibus_config_gconf_new (GDBusConnection *connection)
 {
diff --git a/ibus/config.py b/ibus/config.py
index 0f6e80f..49649bd 100644
--- a/ibus/config.py
+++ b/ibus/config.py
@@ -49,6 +49,9 @@ class ConfigBase(object.Object):
     def unset_value(self, section, name):
         pass
 
+    def get_unused(self, unread, unwritten):
+        pass
+
     def value_changed(self, section, name, value):
         self.__proxy.ValueChanged(section, name, value)
 
@@ -68,6 +71,9 @@ class ConfigProxy(interface.IConfig):
     def UnsetValue(self, section, name):
         return self.__config.unset_value(section, name)
 
+    def GetUnused(self, unread, unwritten):
+        return self.__config.get_unused(unread, unwritten)
+
     def Destroy(self):
         self.__config.destroy()
 
@@ -145,6 +151,12 @@ class Config(object.Object):
         except:
             return
 
+    def get_unused(self, unread, unwritten):
+        try:
+            return self.__config.GetUnused(unread, unwritten)
+        except:
+            return
+
     def set_list(self, section, name, value, signature):
         return self.set_value(section, name, dbus.Array(value, signature=signature))
 
diff --git a/memconf/config.c b/memconf/config.c
index a5aa190..ba738c4 100644
--- a/memconf/config.c
+++ b/memconf/config.c
@@ -29,6 +29,8 @@ typedef struct _IBusConfigMemconfClass IBusConfigMemconfClass;
 struct _IBusConfigMemconf {
     IBusConfigService parent;
     GHashTable *values;
+    GHashTable *unread;
+    GHashTable *unwritten;
 };
 
 struct _IBusConfigMemconfClass {
@@ -52,6 +54,10 @@ static gboolean     ibus_config_memconf_unset_value (IBusConfigService      *con
                                                      const gchar            *section,
                                                      const gchar            *name,
                                                      GError                **error);
+static gboolean     ibus_config_memconf_get_unused  (IBusConfigService      *config,
+                                                     GVariant              **unread,
+                                                     GVariant              **unwritten,
+                                                     GError                **error);
 
 G_DEFINE_TYPE (IBusConfigMemconf, ibus_config_memconf, IBUS_TYPE_CONFIG_SERVICE)
 
@@ -64,6 +70,7 @@ ibus_config_memconf_class_init (IBusConfigMemconfClass *class)
     IBUS_CONFIG_SERVICE_CLASS (object_class)->set_value   = ibus_config_memconf_set_value;
     IBUS_CONFIG_SERVICE_CLASS (object_class)->get_value   = ibus_config_memconf_get_value;
     IBUS_CONFIG_SERVICE_CLASS (object_class)->unset_value = ibus_config_memconf_unset_value;
+    IBUS_CONFIG_SERVICE_CLASS (object_class)->get_unused  = ibus_config_memconf_get_unused;
 }
 
 static void
@@ -71,14 +78,25 @@ ibus_config_memconf_init (IBusConfigMemconf *config)
 {
     config->values = g_hash_table_new_full (g_str_hash,
                                             g_str_equal,
-                                            (GDestroyNotify)g_free,
-                                            (GDestroyNotify)g_variant_unref);
+                                            (GDestroyNotify) g_free,
+                                            (GDestroyNotify) g_variant_unref);
+    config->unread = g_hash_table_new_full (g_str_hash,
+                                            g_str_equal,
+                                            (GDestroyNotify) g_free,
+                                            (GDestroyNotify) NULL);
+    config->unwritten
+                   = g_hash_table_new_full (g_str_hash,
+                                            g_str_equal,
+                                            (GDestroyNotify) g_free,
+                                            (GDestroyNotify) NULL);
 }
 
 static void
 ibus_config_memconf_destroy (IBusConfigMemconf *config)
 {
     g_hash_table_destroy (config->values);
+    g_hash_table_destroy (config->unread);
+    g_hash_table_destroy (config->unwritten);
     IBUS_OBJECT_CLASS (ibus_config_memconf_parent_class)->destroy ((IBusObject *)config);
 }
 
@@ -95,7 +113,14 @@ ibus_config_memconf_set_value (IBusConfigService *config,
     g_assert (value);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
+
+
+    if (g_hash_table_lookup (
+            IBUS_CONFIG_MEMCONF (config)->values, key) == NULL) {
+        g_hash_table_insert (IBUS_CONFIG_MEMCONF (config)->unread,
+                             g_strdup (key), NULL);
+    }
 
     g_hash_table_insert (IBUS_CONFIG_MEMCONF (config)->values,
                          key, g_variant_ref_sink (value));
@@ -116,17 +141,22 @@ ibus_config_memconf_get_value (IBusConfigService *config,
     g_assert (name);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
+
     GVariant *value = (GVariant *)g_hash_table_lookup (IBUS_CONFIG_MEMCONF (config)->values, key);
-    g_free (key);
 
-    if (value != NULL) {
+    if (value == NULL) {
+        g_hash_table_insert (IBUS_CONFIG_MEMCONF (config)->unwritten,
+                             g_strdup (key), NULL);
+        if (error != NULL) {
+            *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                  "Config value [%s:%s] does not exist.", section, name);
+        }
+    } else {
+        g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->unread, key);
         g_variant_ref (value);
     }
-    else if (error != NULL) {
-        *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
-                              "Config value [%s:%s] does not exist.", section, name);
-    }
+    g_free (key);
     return value;
 }
 
@@ -141,11 +171,11 @@ ibus_config_memconf_unset_value (IBusConfigService *config,
     g_assert (name);
     g_assert (error == NULL || *error == NULL);
 
-    gchar *key = g_strdup_printf ("%s:%s", section, name);
+    gchar *key = g_strdup_printf ("%s%s", section, name);
     gboolean retval = g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->values, key);
-    g_free (key);
 
     if (retval) {
+        g_hash_table_remove (IBUS_CONFIG_MEMCONF (config)->unread, key);
     }
     else {
         if (error && *error) {
@@ -153,9 +183,43 @@ ibus_config_memconf_unset_value (IBusConfigService *config,
                                   "Config value [%s:%s] does not exist.", section, name);
         }
     }
+    g_free (key);
     return retval;
 }
 
+static gboolean
+ibus_config_memconf_get_unused (IBusConfigService      *config,
+                                GVariant              **unread,
+                                GVariant              **unwritten,
+                                GError                **error)
+{
+    GList *keys;
+    GList *p;
+    GVariantBuilder builder;
+
+    g_variant_builder_init (&builder, G_VARIANT_TYPE ("as"));
+
+    keys = g_hash_table_get_keys (IBUS_CONFIG_MEMCONF (config)->unread);
+    for (p = keys; p != NULL; p = p->next) {
+        g_variant_builder_add (&builder, "s", (const gchar *)p->data);
+    }
+    g_list_free (keys);
+
+    *unread = g_variant_builder_end (&builder);
+
+    g_variant_builder_init (&builder, G_VARIANT_TYPE ("as"));
+
+    keys = g_hash_table_get_keys (IBUS_CONFIG_MEMCONF (config)->unwritten);
+    for (p = keys; p != NULL; p = p->next) {
+        g_variant_builder_add (&builder, "s", (const gchar *)p->data);
+    }
+    g_list_free (keys);
+
+    *unwritten = g_variant_builder_end (&builder);
+
+    return TRUE;
+}
+
 IBusConfigMemconf *
 ibus_config_memconf_new (GDBusConnection *connection)
 {
diff --git a/src/ibusconfig.c b/src/ibusconfig.c
index 1178ad7..62c434e 100644
--- a/src/ibusconfig.c
+++ b/src/ibusconfig.c
@@ -296,3 +296,37 @@ ibus_config_unset (IBusConfig   *config,
     g_variant_unref (result);
     return TRUE;
 }
+
+gboolean
+ibus_config_get_unused (IBusConfig  *config,
+                        GVariant   **unread,
+                        GVariant   **unwritten)
+{
+    g_assert (IBUS_IS_CONFIG (config));
+    g_assert (unread != NULL && *unread == NULL);
+    g_assert (unwritten != NULL && *unwritten == NULL);
+
+    GError *error = NULL;
+    GVariant *result;
+
+    result = g_dbus_proxy_call_sync ((GDBusProxy *) config,
+                                     "GetUnused",               /* method_name */
+                                     NULL,                      /* parameters */
+                                     G_DBUS_CALL_FLAGS_NONE,    /* flags */
+                                     -1,                        /* timeout */
+                                     NULL,                      /* cancellable */
+                                     &error                     /* error */
+                                     );
+    if (result == NULL) {
+        g_warning ("%s.GetUnused: %s", IBUS_INTERFACE_CONFIG, error->message);
+        g_error_free (error);
+        return FALSE;
+    }
+
+    *unread = g_variant_get_child_value (result, 0);
+    *unwritten = g_variant_get_child_value (result, 1);
+
+    g_variant_unref (result);
+
+    return TRUE;
+}
diff --git a/src/ibusconfig.h b/src/ibusconfig.h
index c3a0bde..196aae9 100644
--- a/src/ibusconfig.h
+++ b/src/ibusconfig.h
@@ -213,6 +213,20 @@ gboolean         ibus_config_set_value_async_finish
 gboolean         ibus_config_unset      (IBusConfig         *config,
                                          const gchar        *section,
                                          const gchar        *name);
+
+/**
+ * ibus_config_get_unused:
+ * @config: An IBusConfig
+ * @unread: GVariant that holds a list of values that have been written but not
+ * read.
+ * @unwritten: GVariant that holds a list of values that have been read but not
+ * written.
+ * @returns: TRUE if succeed; FALSE otherwise.
+ *
+ * Get the list of values that haven't been used properly.
+ */
+gboolean         ibus_config_get_unused     (IBusConfig         *config,
+                                             GVariant          **unread,
+                                             GVariant          **unwritten);
 G_END_DECLS
 #endif
-
diff --git a/src/ibusconfigservice.c b/src/ibusconfigservice.c
index 5374a76..11f75e4 100644
--- a/src/ibusconfigservice.c
+++ b/src/ibusconfigservice.c
@@ -83,6 +83,10 @@ static gboolean  ibus_config_service_unset_value     (IBusConfigService      *co
                                                       const gchar            *section,
                                                       const gchar            *name,
                                                       GError                **error);
+static gboolean ibus_config_service_get_unused       (IBusConfigService      *config,
+                                                      GVariant              **unread,
+                                                      GVariant              **unwritten,
+                                                      GError                **error);
 
 G_DEFINE_TYPE (IBusConfigService, ibus_config_service, IBUS_TYPE_SERVICE)
 
@@ -103,6 +107,10 @@ static const gchar introspection_xml[] =
     "      <arg direction='in'  type='s' name='section' />"
     "      <arg direction='in'  type='s' name='name' />"
     "    </method>"
+    "    <method name='GetUnused'>"
+    "      <arg direction='out'  type='as' name='unread' />"
+    "      <arg direction='out'  type='as' name='unwritten' />"
+    "    </method>"
     "  </interface>"
     "</node>";
 
@@ -125,6 +133,7 @@ ibus_config_service_class_init (IBusConfigServiceClass *class)
     class->set_value   = ibus_config_service_set_value;
     class->get_value   = ibus_config_service_get_value;
     class->unset_value = ibus_config_service_unset_value;
+    class->get_unused  = ibus_config_service_get_unused;
 }
 
 static void
@@ -249,6 +258,25 @@ ibus_config_service_service_method_call (IBusService           *service,
         return;
     }
 
+    if (g_strcmp0 (method_name, "GetUnused") == 0) {
+        GVariant *unread = NULL;
+        GVariant *unwritten = NULL;
+        gboolean retval;
+        GError *error = NULL;
+
+        retval = IBUS_CONFIG_SERVICE_GET_CLASS (config)->get_unused (
+                        config, &unread, &unwritten, &error);
+        if (retval) {
+            g_dbus_method_invocation_return_value (invocation,
+                            g_variant_new ("(@as@as)", unread, unwritten));
+        }
+        else {
+            g_dbus_method_invocation_return_gerror (invocation, error);
+            g_error_free (error);
+        }
+        return;
+    }
+
     /* should not be reached */
     g_return_if_reached ();
 }
@@ -347,6 +375,19 @@ ibus_config_service_new (GDBusConnection *connection)
     return IBUS_CONFIG_SERVICE (object);
 }
 
+static gboolean
+ibus_config_service_get_unused (IBusConfigService  *config,
+                                GVariant          **unread,
+                                GVariant          **unwritten,
+                                GError            **error)
+{
+    if (error) {
+        *error = g_error_new (G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                              "Not implemented");
+    }
+    return FALSE;
+}
+
 void
 ibus_config_service_value_changed (IBusConfigService  *config,
                                    const gchar        *section,
diff --git a/src/ibusconfigservice.h b/src/ibusconfigservice.h
index 1aab32f..c6c62ae 100644
--- a/src/ibusconfigservice.h
+++ b/src/ibusconfigservice.h
@@ -200,10 +200,14 @@ struct _IBusConfigServiceClass {
                                  const gchar          *section,
                                  const gchar          *name,
                                  GError              **error);
+    gboolean    (* get_unused)  (IBusConfigService    *config,
+                                 GVariant            **unread,
+                                 GVariant            **unwritten,
+                                 GError              **error);
 
     /*< private >*/
     /* padding */
-    gpointer pdummy[13];
+    gpointer pdummy[12];
 };
 
 GType                ibus_config_service_get_type   (void);
@@ -235,4 +239,3 @@ void                 ibus_config_service_value_changed
 
 G_END_DECLS
 #endif
-
-- 
1.7.3.1

